vowpalwabbit/parse_example.cc

```diff
@@ -50,29 +50,114 @@ int read_features_string(vw* all, v_array<example*>& examples)
+  features& get_affix_fg() { return get_feature_group(_p_affix_feature_group, affix_namespace); }
+  features& get_spelling_fg() { return get_feature_group(_p_spelling_feature_group, spelling_namespace); }
+  features& get_dictionary_fg() { return get_feature_group(_p_dictionary_feature_group, dictionary_namespace); }
+  void set_current_fg(const VW::string_view fg_name)
+  {
+    _feature_group_name = fg_name;
+    auto namespace_char = static_cast<unsigned char>(_feature_group_name[0]);
+    if (_redefine_char_flag) namespace_char = (*_redefine_char_map)[namespace_char];  // redefine feature_group_char
+    _feature_group_char = namespace_char;
+    _hash = _hash_function(_feature_group_name.begin(), _feature_group_name.length(), _hash_seed);
+    _p_current_feature_group = nullptr;  // reset cached feature group
+    _affix = (*_affix_features)[_feature_group_char];
+    _spelling = (*_spelling_features)[_feature_group_char];
+    _namespace_dictionaries = &(*_namespace_dictionaries_map)[feature_group_char()];
+  }
+  void set_current_fg()
+  {
+    _feature_group_name = _default_namespace_str;
+    _feature_group_char = default_namespace;
+    _hash = this->_hash_seed == 0 ? 0 : uniform_hash("", 0, this->_hash_seed);
+    _p_current_feature_group = nullptr;  // reset cached feature group
+    _affix = (*_affix_features)[_feature_group_char];
+    _spelling = (*_spelling_features)[_feature_group_char];
+    _namespace_dictionaries = &(*_namespace_dictionaries_map)[feature_group_char()];
+  }
+  const VW::string_view get_fg_name() const { return _feature_group_name; }
+  uint64_t get_affix_features() const { return _affix; }
+  bool is_spelling_feature_group() const { return _spelling; }
+  bool dictionary_exists() const { return !_namespace_dictionaries->empty(); }
+  const namespace_dictionaries* get_dictionary() const { return _namespace_dictionaries; }
+
+private:
+  features& get_feature_group(features*& p_features_ref, unsigned char namespace_char)
+  {
+    if (p_features_ref == nullptr)
+    {
+      p_features_ref = &_current_example->feature_space[namespace_char];
+
+      if (p_features_ref->empty()) _current_example->indices.emplace_back(namespace_char);
+    }
+    return *p_features_ref;
+  }
+
+  unsigned char _feature_group_char;
+  int64_t _hash;
+  uint32_t _hash_seed;
+  features* _p_current_feature_group;
+  features* _p_affix_feature_group;
+  features* _p_spelling_feature_group;
+  features* _p_dictionary_feature_group;
+  example* _current_example;
+  bool _redefine_char_flag;
+  VW::string_view _feature_group_name;
+  uint64_t _affix;
+  bool _spelling;
+  namespace_dictionaries* _namespace_dictionaries;
+  char_map* _redefine_char_map;
+  dictionaries_map* _namespace_dictionaries_map;
+  affix_map* _affix_features;
+  bool_map* _spelling_features;
+  hash_func_t _hash_function;
+  const char* _default_namespace_str = " ";
```

Comment:
Why make this is a member? Seems reasonable to make it a const global?

Reply:
yeah makes sense.